### PR TITLE
464/lru cache

### PR DESCRIPTION
**Pull request by kasbaker**
# Overview:
This pull request simply adds the `@lru_cache` decorator to `EphysNWBData._get_series()` in order to improve performance.

# Addresses:
Addresses issue: https://github.com/AllenInstitute/ipfx/issues/464

# Type of Fix:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)
- [ ] Documentation Change

# Solution:
The function `_get_series()` is being called multiple times when trying to access the same data, which slows down the auto QC pipeline significantly. Adding this decorator greatly reduces the amount of time it takes to call this function after the initial call.

# Changes:
- Added `from functools import lru_cache` to `ephys_nwb_data.py`.
- Added `@lru_cache(maxsize=None)` to the `_get_series()` method of `EphysNWBData`.
- Updated `CHANGELOG.md`.
- Updated `version.txt`.

# Validation:
I ran `cProfile` on a function that ran cell and sweep qc on two different versions of `EphysDataSet`. One had the `@lru_cache` decorator and the other did not. Screenshots below show the performace increase. More profile files are available if needed.

### Screenshots:
Snakeviz screenshots before and after adding `@lru_cache` decorator:

image
Before using LRU cache decorator

image
After using LRU cache decorator
### Unit Tests:
This did not change anything significantly enough to require new unit tests.
### Script to reproduce error and fix:
``` python
from pathlib import Path
from copy import deepcopy
import datetime as dt
import cProfile
import pstats
import json
import functools
import gc

from ipfx.dataset.create import create_ephys_data_set, create_ephys_data_set_lru
from ipfx.sweep_props import drop_tagged_sweeps
from ipfx.bin.run_qc import qc_summary
from ipfx.qc_feature_extractor import cell_qc_features, sweep_qc_features
from ipfx.qc_feature_evaluator import qc_experiment, DEFAULT_QC_CRITERIA_FILE
from ipfx.stimulus import StimulusOntology


with open(DEFAULT_QC_CRITERIA_FILE, "r") as path:
    QC_CRITERIA = json.load(path)

with open(StimulusOntology.DEFAULT_STIMULUS_ONTOLOGY_FILE, "r") \
        as path:
    ONTOLOGY = StimulusOntology(json.load(path))


def main(nwb_file, lru = False):
    if lru:
        print("testing LRU cache")
        data_set = create_ephys_data_set_lru(nwb_file=nwb_file, ontology=ONTOLOGY)
    else:
        print("testing no cache")
        data_set = create_ephys_data_set(nwb_file=nwb_file, ontology=ONTOLOGY)

    # cell QC worker
    cell_features, cell_tags = cell_qc_features(data_set)
    cell_features = deepcopy(cell_features)

    # sweep QC worker
    sweep_features = sweep_qc_features(data_set)
    sweep_features = deepcopy(sweep_features)
    drop_tagged_sweeps(sweep_features)

    # experiment QC worker
    cell_state, sweep_states = qc_experiment(
        ontology=ONTOLOGY,
        cell_features=cell_features,
        sweep_features=sweep_features,
        qc_criteria=QC_CRITERIA
    )

    qc_summary(
        sweep_features=sweep_features,
        sweep_states=sweep_states,
        cell_features=cell_features,
        cell_state=cell_state
    )


def clear_all_lru_cache():
    # All objects collected
    objects = [i for i in gc.get_objects()
               if isinstance(i, functools._lru_cache_wrapper)]

    # All objects cleared
    for object in objects:
        object.cache_clear()


if __name__ == '__main__':
    files = list(Path("data/nwb").glob("*.nwb"))
    base_dir = Path(__file__).parent
    today = dt.datetime.now().strftime('%y%m%d')

    now = dt.datetime.now().strftime('%H.%M.%S')
    profile_dir = base_dir.joinpath(f'profiles/{today}/{now}')
    profile_dir.mkdir(parents=True)
    for file in files:
        nwb_file = str(file)

        profile_file = str(profile_dir.joinpath(f'lru_cache_get_series_no_max-{str(file.stem)}.prof'))
        cProfile.run('main(nwb_file, lru=True)', filename=profile_file)
        p = pstats.Stats(profile_file)
        p.sort_stats('cumtime').print_stats(20)

        clear_all_lru_cache()

        profile_file = str(profile_dir.joinpath(f'no_cache_{str(file.stem)}.prof'))
        cProfile.run('main(nwb_file, lru=False)', filename=profile_file)
        p = pstats.Stats(profile_file)
        p.sort_stats('cumtime').print_stats(20)

        clear_all_lru_cache()
```
### Configuration details:

# Checklist
- [x] My code follows
      [Allen Institute Contribution Guidelines](CONTRIBUTING.md)
- [x] My code is unit tested and does not decrease test coverage
- [x] I have performed a self review of my own code
- [x] My code is well-documented, and the docstrings conform to
      [Numpy Standards](https://numpydoc.readthedocs.io/en/latest/format.html)
- [x] I have updated the documentation of the repository where
      appropriate
- [x] The header on my commit includes the issue number
- [x] My code passes all tests
- [x] I have updated the CHANGELOG.md with the description of changes understandable by end users

# Notes:
- ~~There are some issues with the build. Installing with `--use-feature=2020-resolver` might help with some of the dependency issues.~~ Used `pip freeze > requirements.txt` on my machine to fix requirements. Build works now, but all requirements are pinned.
- ~~Some pytests failed this branch, but the same ones also failed on the master branch.~~ All tests are passing on CircleCi when the server runs them.

### CODE REVIEW

**Review by MatthewAitken — COMMENTED** [review]
lgtm, nice work. Since I'm not as familiar with IPFX, I asked Sergey to take a look as well before we merge it.

**Review by sgratiy — CHANGES_REQUESTED** [review]
Looks great as far as LRU cache.
Some small things:
I have changed  the base to "dev" . Only hotfixes get to be merged to master, otherwise we merge into dev. Having said that we just recently had a hotfix and already released 1.0.1. 
Unfortunately, one of our team did not merged master with the hotfix back into dev.
Please merge master into dev, pull latest dev and rebase your work on top of it.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,10 +1,15 @@
 # Change Log
 All notable changes to this project will be documented in this file.
 
+## [1.0.1] = 2020-09-22
```
Comment (sgratiy) [inline CHANGELOG.md:4]: We already have 1.0.1 but unfortunately we did not keep the change log for that hotfix.
Per CONTRIBUTION_GUIDELINES.md:
Update the "Unreleased" section in the CHANGELOG.md with the change made in this PR.

**Review by sgratiy — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (kasbaker) [reply CHANGELOG.md]: I moved the changes to "Unreleased" under the headers "Added" and "Changed".

**Comment by kasbaker** [comment]
> Looks great as far as LRU cache.
> Some small things:
> I have changed the base to "dev" . Only hotfixes get to be merged to master, otherwise we merge into dev. Having said that we just recently had a hotfix and already released 1.0.1.
> Unfortunately, one of our team did not merged master with the hotfix back into dev.
> Please merge master into dev, pull latest dev and rebase your work on top of it.

@sgratiy I just did this, but it doesn't seem to have changed anything because upstream master is 26 commits ahead of upstream dev. Do you want me to change this pull request to a pull request for the dev branch?

**Comment by sgratiy** [comment]
@kasbaker 
As a contributor you should always target dev branch in your PRs
I have looked at our history and after fetching I see the following:
<pre>* <font color="#C4A000">0fc0a74 (</font><font color="#EF2929"><b>origin/master</b></font><font color="#C4A000">, </font><font color="#EF2929"><b>origin/HEAD</b></font><font color="#C4A000">, </font><font color="#8AE234"><b>master</b></font><font color="#C4A000">)</font> Added additional instruction for clarification
* <font color="#C4A000">b896302 (</font><font color="#FCE94F"><b>tag: v1.0.1</b></font><font color="#C4A000">)</font> Updated version
* <font color="#C4A000">d430f12</font> Bumped version to 1.0.1
* <font color="#C4A000">17eeafc</font> GH-471 Make dedicated docker image for bamboo CI
* <font color="#C4A000">43f6f69</font> GH-471 Update pynwb requirements
*   <font color="#C4A000">995f371 (</font><font color="#FCE94F"><b>tag: v1.0.0</b></font><font color="#C4A000">)</font> Merge pull request #458 from AllenInstitute/release
<font color="#4E9A06">|</font><font color="#C4A000">\</font>  
<font color="#4E9A06">|</font> * <font color="#C4A000">4eaf4f9 (</font><font color="#EF2929"><b>origin/release</b></font><font color="#C4A000">)</font> 1.0.0 updates
<font color="#4E9A06">|</font> *   <font color="#C4A000">e681009</font> Merge branch &apos;dev&apos; into release
<font color="#4E9A06">|</font> <font color="#3465A4">|</font><font color="#75507B">\</font>  
<font color="#4E9A06">|</font> <font color="#3465A4">|</font> * <font color="#C4A000">e8545c0 (</font><font color="#34E2E2"><b>HEAD -&gt; </b></font><font color="#8AE234"><b>onto-tag</b></font><font color="#C4A000">, </font><font color="#EF2929"><b>origin/dev</b></font><font color="#C4A000">, </font><font color="#8AE234"><b>dev</b></font><font color="#C4A000">)</font> 428 Implements graceful degradation for feature extractor  (#449)
<font color="#4E9A06">|</font> * <font color="#75507B">|</font>   <font color="#C4A000">68355e7</font> Merge pull request #457 from AllenInstitute/dev
<font color="#4E9A06">|</font> <font color="#06989A">|</font><font color="#75507B">\</font> <font color="#75507B">\</font>  
<font color="#4E9A06">|</font> <font color="#06989A">|</font> <font color="#75507B">|/</font>  
</pre>

I do not see the 26 commits difference. Did you fetch the latest?

**Comment by sgratiy** [comment]
Ok, I see the 26 commits now.

**Comment by sgratiy** [comment]
@kasbaker I have created a separate PR to merge master into dev. Thanks for pointing out that the 26 commit difference.

**Comment by kasbaker** [comment]
> @kasbaker I have created a separate PR to merge master into dev. Thanks for pointing out that the 26 commit difference.

OK, I changed the base branch of this PR from master to dev, so the other PR you created might not be necessary since this was originally based off master.

**Pull request merged**